crypto/pkcs7/pkcs7_test.cc

```diff
@@ -1044,3 +1129,212 @@ hJTbHtjEDJ7BHLC/CNUhXbpyyu1y
+    EXPECT_TRUE(PKCS7_type_is_signed(p7.get()));
+    EXPECT_TRUE(PKCS7_content_new(p7.get(), NID_pkcs7_signed));
+    EXPECT_FALSE(PKCS7_set_cipher(p7.get(), EVP_aes_128_gcm()));
+    EXPECT_FALSE(PKCS7_add_recipient_info(p7.get(), nullptr));
+    EXPECT_FALSE(PKCS7_get_signer_info(nullptr));
+
+    p7.reset(PKCS7_new());
+    ASSERT_TRUE(p7);
+    EXPECT_TRUE(PKCS7_set_type(p7.get(), NID_pkcs7_digest));
+    EXPECT_TRUE(PKCS7_type_is_digest(p7.get()));
+    EXPECT_TRUE(PKCS7_content_new(p7.get(), NID_pkcs7_digest));
+    EXPECT_FALSE(PKCS7_add_certificate(p7.get(), nullptr));
+    EXPECT_FALSE(PKCS7_add_crl(p7.get(), nullptr));
+    EXPECT_FALSE(PKCS7_add_signer(p7.get(), nullptr));
+    EXPECT_FALSE(PKCS7_get_signer_info(p7.get()));
+
+    p7.reset(PKCS7_new());
+    ASSERT_TRUE(p7.get());
+    EXPECT_TRUE(PKCS7_set_type(p7.get(), NID_pkcs7_data));
+    EXPECT_TRUE(PKCS7_type_is_data(p7.get()));
+    EXPECT_FALSE(PKCS7_set_content(p7.get(), p7.get()));
+
+    p7.reset(PKCS7_new());
+    ASSERT_TRUE(p7.get());
+    EXPECT_TRUE(PKCS7_set_type(p7.get(), NID_pkcs7_signedAndEnveloped));
+    EXPECT_TRUE(PKCS7_type_is_signedAndEnveloped(p7.get()));
+    EXPECT_TRUE(PKCS7_set_cipher(p7.get(), EVP_aes_128_gcm()));
+    EXPECT_FALSE(PKCS7_set_content(p7.get(), p7.get()));
+
+    p7.reset(PKCS7_new());
+    ASSERT_TRUE(p7.get());
+    EXPECT_TRUE(PKCS7_set_type(p7.get(), NID_pkcs7_enveloped));
+    EXPECT_TRUE(PKCS7_type_is_enveloped(p7.get()));
+    EXPECT_TRUE(PKCS7_set_cipher(p7.get(), EVP_aes_128_gcm()));
+    EXPECT_FALSE(PKCS7_set_content(p7.get(), p7.get()));
+
+    p7.reset(PKCS7_new());
+    ASSERT_TRUE(p7.get());
+    EXPECT_TRUE(PKCS7_set_type(p7.get(), NID_pkcs7_encrypted));
+    EXPECT_TRUE(PKCS7_type_is_encrypted(p7.get()));
+    EXPECT_FALSE(PKCS7_set_content(p7.get(), p7.get()));
+
+    const uint8_t *p7_der = kPKCS7SignedWithSignerInfo;
+    const size_t p7_der_len = sizeof(kPKCS7SignedWithSignerInfo);
+    p7.reset(d2i_PKCS7(nullptr, &p7_der, p7_der_len));
+    ASSERT_TRUE(p7);
+    ASSERT_TRUE(PKCS7_type_is_signed(p7.get()));
+    STACK_OF(PKCS7_SIGNER_INFO) *sk_p7si_signed = PKCS7_get_signer_info(p7.get());
+    ASSERT_TRUE(sk_p7si_signed);
+    ASSERT_GT(sk_PKCS7_SIGNER_INFO_num(sk_p7si_signed), 0UL);
+    PKCS7_SIGNER_INFO *p7si = sk_PKCS7_SIGNER_INFO_value(sk_p7si_signed, 0);
+    ASSERT_TRUE(p7si);
+    EXPECT_FALSE(PKCS7_get_signed_attribute(p7si, NID_md5));    // hash nid not valid x509 attr
+    ASN1_TYPE *signing_time = PKCS7_get_signed_attribute(p7si, NID_pkcs9_signingTime);
+    ASSERT_TRUE(signing_time);
+    EVP_PKEY *pkey;
+    X509_ALGOR *pdig;
+    X509_ALGOR *psig;
+    PKCS7_SIGNER_INFO_get0_algs(p7si, &pkey, &pdig, &psig);
+    ASSERT_FALSE(pkey); // no attached pkey
+    ASSERT_TRUE(psig);
+    ASSERT_TRUE(pdig);
+
+    bssl::UniquePtr<PKCS7> p7_dup(PKCS7_dup(p7.get()));
+    ASSERT_TRUE(p7_dup);
+    EXPECT_TRUE(PKCS7_type_is_signed(p7_dup.get()));
+
+    p7_der = kPKCS7SignedWithSignerInfo;
```

Comment:
Minor nit: is this necessary, seems like the pointer value hasn't changed.

Reply:
are you asking about the `PKCS7_dup` call on L1207? if so, that's meant to exercise `PKCS7_dup` and assert that a.) `PKCS7_dup` duplicates the object without memory leaks and b.) that the duplicated object is of the expected content type.

if asking about the `p7_der` assignment on L1211, then yes, and nice catch :) it's intentional. `d2i_*` functions will increment the input reference during parsing (see comment [here](https://github.com/aws/aws-lc/blob/f88ef5dff488c3a36f65390ff5a0a8e92e3d941b/crypto/pkcs7/pkcs7_asn1.c#L55)). this is why we assign a local pointer to the static address and reset it for relevant tests. OpenSSL documents the generic behavior in [this type-specific man page](https://docs.openssl.org/1.1.1/man3/d2i_X509).

> d2i_TYPE() attempts to decode len bytes at *ppin. If successful a pointer to the TYPE structure is returned and *ppin is incremented to the byte following the parsed data.

thanks for the call-out, this needs to be documented in a comment.

Reply:
Ah yes I was asking about `p7_der` on L1211. It seemed like the `d2i` functions did some pointer advancements, but the `const` pointer made me question myself if anything had actually changed. Thanks for the clarification!!